homeassistant/components/fronius/sensor.py

```diff
@@ -1,354 +1,226 @@
-        self, bridge: Fronius, name: str, device: int, add_entities: AddEntitiesCallback
+        self,
+        coordinator: _FroniusUpdateCoordinator,
+        entity_description: SensorEntityDescription,
+        solar_net_id: str,
     ) -> None:
-        """Initialize the sensor."""
-        self.bridge = bridge
-        self._name = name
-        self._device = device
-        self._fetched: dict[str, Any] = {}
-        self._available = True
-
-        self.sensors: set[str] = set()
-        self._registered_sensors: set[SensorEntity] = set()
-        self._add_entities = add_entities
-
-    @property
-    def name(self):
-        """Return the name of the sensor."""
-        return self._name
-
-    @property
-    def data(self):
-        """Return the state attributes."""
-        return self._fetched
-
-    @property
-    def available(self):
-        """Whether the fronius device is active."""
-        return self._available
-
-    async def async_update(self):
-        """Retrieve and update latest state."""
-        try:
-            values = await self._update()
-        except ConnectionError:
-            # fronius devices are often powered by self-produced solar energy
-            # and henced turned off at night.
-            # Therefore we will not print multiple errors when connection fails
-            if self._available:
-                self._available = False
-                _LOGGER.error("Failed to update: connection error")
-            return
-        except ValueError:
-            _LOGGER.error(
-                "Failed to update: invalid response returned."
-                "Maybe the configured device is not supported"
-            )
-            return
-
-        self._available = True  # reset connection failure
-
-        attributes = self._fetched
-        # Copy data of current fronius device
-        for key, entry in values.items():
-            # If the data is directly a sensor
-            if "value" in entry:
-                attributes[key] = entry
-        self._fetched = attributes
-
-        # Add discovered value fields as sensors
-        # because some fields are only sent temporarily
-        new_sensors = []
-        for key in attributes:
-            if key not in self.sensors:
-                self.sensors.add(key)
-                _LOGGER.info("Discovered %s, adding as sensor", key)
-                new_sensors.append(FroniusTemplateSensor(self, key))
-        self._add_entities(new_sensors, True)
-
-        # Schedule an update for all included sensors
-        for sensor in self._registered_sensors:
-            sensor.async_schedule_update_ha_state(True)
-
-    async def _update(self) -> dict:
-        """Return values of interest."""
+        """Set up an individual Fronius meter sensor."""
+        super().__init__(coordinator)
+        self.entity_description = entity_description
+        self.solar_net_id = solar_net_id
+        self._attr_native_value = self._get_entity_value()
```

Comment:
Does this work? My understanding is that this will set the native value on init only. I would expect you should just rename the `_get_entity_value` function to `native_value`.

Reply:
Famous last words: It works fine on my system 🤣

The entities are instantiated after instantiating the coordinator, so `_handle_coordinator_update()` is not called after an entity is created and would not have an init value then. Later updates are handled by `_handle_coordinator_update()`

Since we can't be sure if the key is contained in the latest API response (`coordinator.data` dict) the state would turn to `None` instead of stay on the latest known value. This happens eg. after sunset when an inverter doesn't produce anymore - Fronius API just omits some keys then 🙄

Reply:
Actually, the state turning to None is not a problem. This is the intended way of telling HA that the current state could not be fetched. AFAIK HA will still display the latest non-None retrieved value.

Reply:
I've just tested this. At night I would get these API results from inverters - alternating:
```log
[pyfronius] Converting inverter data from '{'DAY_ENERGY': {'Unit': 'Wh', 'Value': 112640}, 'DeviceStatus': {'ErrorCode': 307, 'LEDColor': 1, 'LEDState': 0, 'MgmtTimerRemainingTime': 14, 'StateToReset': False, 'StatusCode': 3}, 'IDC': {'Unit': 'A', 'Value': 0}, 'TOTAL_ENERGY': {'Unit': 'Wh', 'Value': 43421500}, 'UDC': {'Unit': 'V', 'Value': 4.4}, 'YEAR_ENERGY': {'Unit': 'Wh', 'Value': 24742314}}'
[pyfronius] Converting inverter data from '{'DAY_ENERGY': {'Unit': 'Wh', 'Value': 112640}, 'DeviceStatus': {'ErrorCode': 523, 'LEDColor': 3, 'LEDState': 0, 'MgmtTimerRemainingTime': -1, 'StateToReset': True, 'StatusCode': 2}, 'TOTAL_ENERGY': {'Unit': 'Wh', 'Value': 43421500}, 'YEAR_ENERGY': {'Unit': 'Wh', 'Value': 24742314}}'
```
ErrorCode `307` contains Current and Voltage, `523` does not. 
With a `native_value` property returning `None` if the key was not in the result the state would alternate between `Unknown` and `0` every minute or so (its not regularly alternating).

Reply:
Well I figured out where I was misunderstanding this, you are replacing the `_handle_coordinator_update` callback to update `_attr_native_value`.